hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala

```diff
@@ -915,13 +939,17 @@ object EmitStream {
                       cb.assign(rightProducer.elementRegion, Region.stagedCreate(Region.REGULAR, outerRegion.getPool()))
                     else
                       cb.assign(rightProducer.elementRegion, outerRegion)
+                    if (leftProducer.requiresMemoryManagementPerElement)
+                      cb.assign(leftProducer.elementRegion, Region.stagedCreate(Region.REGULAR, outerRegion.getPool()))
+                    else
+                      cb.assign(leftProducer.elementRegion, outerRegion)
 
                     leftProducer.initialize(cb)
                     rightProducer.initialize(cb)
                   }
 
-                  override val elementRegion: Settable[Region] = leftProducer.elementRegion
-                  override val requiresMemoryManagementPerElement: Boolean = leftProducer.requiresMemoryManagementPerElement
+                  override val elementRegion: Settable[Region] = _elementRegion
```

Comment:
the old implementation could generate region reference cycles if the left stream didn't require memory management, and the right did.